src/libstd/ffi/os_str.rs

```diff
@@ -83,6 +85,33 @@ impl OsString {
         OsString { inner: Buf::from_string(String::new()) }
     }
 
+    /// Construct an `OsString` from a byte sequence.
+    ///
+    /// # Platform behavior
+    ///
+    /// On Unix systems, any byte sequence can be successfully
+    /// converted into an `OsString`.
+    ///
+    /// On Windows system, only UTF-8 byte sequences will successfully
+    /// convert; non UTF-8 data will produce `None`.
+    #[unstable(feature = "convert", reason = "recently added")]
+    pub fn from_bytes<B>(bytes: B) -> Option<OsString> where B: Into<Vec<u8>> {
+        #[cfg(unix)]
+        use os::unix::ffi::OsStringExt;
+
+        if cfg!(windows) {
+            String::from_utf8(bytes.into()).ok().map(OsString::from)
+        } else {
+            Some(OsString::from_vec(bytes.into()))
```

Comment:
Unfortunately I don't think this will compile on windows as `OsString::from_vec` won't resolve otherwise